core/rovecomm.py

```diff
@@ -141,46 +277,187 @@ def read(self):
             return_packet (RoveCommPacket): A RoveCommPacket that contains a
             RoveComm message received over the network
         '''
-        try:
-            packet, remote_ip = self.RoveCommSocket.recvfrom(1024)
-            header_size = struct.calcsize(ROVECOMM_HEADER_FORMAT)
+        # The select function is used to poll the socket and check whether
+        # there is data available to be read, preventing the read from
+        # blocking the thread while waiting for a packet
+        available_sockets = select.select([self.RoveCommSocket], [], [], 0)[0]
+        if len(available_sockets) > 0:
+            try:
+                packet, remote_ip = self.RoveCommSocket.recvfrom(1024)
+                header_size = struct.calcsize(ROVECOMM_HEADER_FORMAT)
 
-            rovecomm_version, data_id, data_count, data_type = struct.unpack(
-                ROVECOMM_HEADER_FORMAT, packet[0:header_size])
-            data = packet[header_size:]
+                rovecomm_version, data_id, data_count, data_type = struct.unpack(
+                    ROVECOMM_HEADER_FORMAT, packet[0:header_size])
+                data = packet[header_size:]
 
-            if rovecomm_version != 2:
-                return_packet = RoveCommPacket(ROVECOMM_INCOMPATIBLE_VERSION, 'b', (1,), '')
+                if rovecomm_version != 2:
+                    return_packet = RoveCommPacket(ROVECOMM_INCOMPATIBLE_VERSION, 'b', (1,), '')
+                    return_packet.ip_address = remote_ip
+                    return return_packet
+
+                if data_id == ROVECOMM_SUBSCRIBE_REQUEST:
+                    if self.subscribers.count(remote_ip) == 0:
+                        self.subscribers.append(remote_ip)
+                elif data_id == ROVECOMM_UNSUBSCRIBE_REQUEST:
+                    if self.subscribers.count(remote_ip) != 0:
+                        self.subscribers.remove(remote_ip)
+
+                data_type = types_int_to_byte[data_type]
+                data = struct.unpack('>' + data_type * data_count, data)
+
+                return_packet = RoveCommPacket(data_id, data_type, data, '')
                 return_packet.ip_address = remote_ip
                 return return_packet
 
-            if data_id == ROVECOMM_SUBSCRIBE_REQUEST:
-                if self.subscribers.count(remote_ip == 0):
-                    self.subscribers.append(remote_ip)
-            elif data_id == ROVECOMM_UNSUBSCRIBE_REQUEST:
-                if self.subscribers.count(remote_ip) != 0:
-                    self.subscribers.remove(remote_ip)
+            except Exception:
+                return_packet = RoveCommPacket()
+                return return_packet
+
+    def close_socket(self):
+        '''
+        Closes the UDP socket
+        '''
+        self.RoveCommSocket.shutdown(1)
```

Comment:
We shouldn't be calling shutdown() on a DGRAM socket. I don't think it has defined behaviour. Shouldn't be a problem to take out.